### DISCUSSION TITLE
How to run a standalone TypeScript type checker on civet files?

### DISCUSSION THREAD

**Discussion by greghuc** [DanielXMoore/Civet#481] [Q&A] [answered]
Hi there

Any advice on how to run a standalone TypeScript type checker on a civet webapp project?

In pure Typescript world, the [recommendation](https://stackoverflow.com/questions/55002137/typescript-noemit-use-case) seems to be to run `tsc --noEmit`. However, I haven't figured how to pass in the civet->ts transpiler into `tsc`.

I have got some nice type errors in civet files with Visual Studio + [Civet language server](https://marketplace.visualstudio.com/items?itemName=DanielX.civet)  + tsconfig.json as the code editor.

But I wanted a standalone type checker to augment my use of esbuild  + [Civet esbuild plugin](https://github.com/DanielXMoore/Civet/blob/main/source/esbuild-plugin.civet) as the webapp build tool. The civet plugin transpiles to js and doesn't seem to error on type issues (and esbuild doesn't process typescript types any way).

Many thanks

Greg

**Comment by edemaine** [comment]
Eventually we plan to have a standalone tool for this: #61

In an esbuild context, I wonder if we could use https://www.npmjs.com/package/@jgoz/esbuild-plugin-typecheck together with #448. You're right that the esbuild plugin currently hard-codes `js: true`, but we could turn that into an option...

**Reply by greghuc** [reply]
@edemaine thanks for the reply. I'm glad I wasn't missing anything obvious. [esbuild-plugin-typecheck](https://www.npmjs.com/package/@jgoz/esbuild-plugin-typecheck) does look promising. However I also like the current simplicity of civet outputting js directly. Otherwise, we might be looking at multiple esbuild plugins for (civet -> ts) + (ts -> js) + (ts -> ts-type-check running in parallel)..

**Reply by greghuc** [reply]
Also, given I have coffeescript in the mix, I'm not too sure how messy things look if everything is assumed to to be typescript, where I would have a mix of ts (from civet) and js (from coffee).

**Reply by edemaine** [reply, marked as answer]
An update on this: The CLI tool can now do typechecking of Civet code via `civet --typecheck`. See https://civet.dev/getting-started#typechecking
I think it will work with imported js files (or ts files) but let us know if you run into trouble.
